### PR TITLE
Add `getOrCreateInstance` method in base-component

### PR DESCRIPTION
**Pull request by GeoSot**
```js
   let data = Data.get(element, DATA_KEY)
  

      if (!data) {
        data = new Alert(element)
      }
```

can be

````js
  const data = Alert.getOrCreateInstance(element)
  ````

### CODE REVIEW

File: js/src/base-component.js [hunk 1]
```diff
@@ -38,6 +38,12 @@ class BaseComponent {
     return Data.get(element, this.DATA_KEY)
   }
 
+  static getInstanceOrNew(element, config = {}) {
```
Comment (rohit2sharma95) [inline js/src/base-component.js:41]: Better name => `getOrCreateInstance` :thinking:

File: js/src/tooltip.js [hunk 2]
```diff
@@ -7,21 +7,8 @@
...
-  isRTL,
-  noop,
-  typeCheckConfig
-} from './util/index'
-import {
-  DefaultAllowlist,
-  sanitizeHtml
-} from './util/sanitizer'
+import { defineJQueryPlugin, emulateTransitionEnd, findShadowRoot, getTransitionDurationFromElement, getUID, isElement, isRTL, noop, typeCheckConfig } from './util/index'
+import { DefaultAllowlist, sanitizeHtml } from './util/sanitizer'
```
Comment (rohit2sharma95) [inline js/src/tooltip.js:11]: Unrelated changes :slightly_smiling_face:

**Review by rohit2sharma95 — COMMENTED** [review]
There is commented code also :grin:

File: js/src/base-component.js (hunk 1 above)
Comment (GeoSot) [reply js/src/base-component.js:41]: I am not sure which is best, but it is very easy to rename, if we want it

**Comment by GeoSot** [comment]
Commented code is gone, unrelated reverted, method is renamed :)

**Comment by XhmikosR** [comment]
Would this conflict with #33249 and #33250? If so, unsure which should land first.

BTW we don't really test jQuery, just if it works, so I'm a little hesitant :/

**Comment by GeoSot** [comment]
The other two, are small easy refactorings , so you can land them first.

File: js/src/collapse.js [hunk 3]
```diff
@@ -332,14 +332,14 @@ class Collapse extends BaseComponent {
...
-    let data = Data.get(element, DATA_KEY)
+    let data = Collapse.getInstance(element)
     const _config = {
       ...Default,
       ...Manipulator.getDataAttributes(element),
       ...(typeof config === 'object' && config ? config : {})
     }
 
-    if (!data && _config.toggle && typeof config === 'string' && /show|hide/.test(config)) {
+    if (!data && _config.toggle && typeof config === 'string' && ['show', 'hide'].includes(config)) {
```
Comment (XhmikosR) [inline js/src/collapse.js:342]: Please revert this an any similar (regex to includes or equal to includes) changes in your PRs. IIRC the check is inverted now. Better keep PRs clean, one change at a time :)

Comment (GeoSot) [reply js/src/collapse.js:342]: I can revert it as it is not on the scope of my Pr, and you are right,, but I am not sure if it is right practice. we can discuss it in another PR

File: js/src/popover.js [hunk 4]
```diff
@@ -136,25 +135,25 @@ class Popover extends Tooltip {
...
-
-      if (!data && /dispose|hide/.test(config)) {
-        return
-      }
+      let data = Popover.getInstance(this)
 
       if (!data) {
-        data = new Popover(this, _config)
-        Data.set(this, DATA_KEY, data)
+        if (typeof config === 'string' && ['dispose', 'hide'].includes(config)) {
```
Comment (XhmikosR) [inline js/src/popover.js:141]: here too

File: js/src/tooltip.js [hunk 5]
```diff
@@ -761,24 +761,21 @@ class Tooltip extends BaseComponent {
...
+      const data = Tooltip.getOrCreateInstance(this, config)
 
-      if (!data && /dispose|hide/.test(config)) {
+      if (typeof config !== 'string') {
         return
       }
 
-      if (!data) {
-        data = new Tooltip(this, _config)
+      if (['dispose', 'hide'].includes(config)) {
```
Comment (XhmikosR) [inline js/src/tooltip.js:770]: And here

**Comment by alecpl** [comment]
It would be great if all these @GeoSot's refactoring PRs could make it to 5.0.0. @mdo any chance for that?

@GeoSot is this still a draft? looks like some tests failed.

**Comment by XhmikosR** [comment]
I'd rather not get the further refactoring PRs in v5.0.0 to not risk to break something so close to the release. There's plenty of time in 5.0.1 or 5.1.0.

**Comment by GeoSot** [comment]
stable 5.0.0 is the most important thing for now. The next thing, I bet are some of the refactorings

File: js/src/dropdown.js [hunk 6]
```diff
@@ -401,7 +395,7 @@ class Dropdown extends BaseComponent {
     const toggles = SelectorEngine.find(SELECTOR_DATA_TOGGLE)
 
     for (let i = 0, len = toggles.length; i < len; i++) {
-      const context = Data.get(toggles[i], DATA_KEY)
+      const context = Dropdown.getInstance(toggles[i], DATA_KEY)
```
Comment (rohit2sharma95) [inline js/src/dropdown.js:398]: The second arguments is not needed :thinking:

File: js/src/carousel.js [hunk 7]
```diff
@@ -219,7 +218,8 @@ class Carousel extends BaseComponent {
   _getConfig(config) {
     config = {
       ...Default,
-      ...config
+      ...Manipulator.getDataAttributes(this._element),
+      ...(typeof config === 'object' ? config : {})
```
Comment (rohit2sharma95) [inline js/src/carousel.js:222]: This check can be omitted here since you have already placed this in the definition of `getOrCreateInstance`?

Comment (GeoSot) [reply js/src/carousel.js:222]: And in case the Component is initialized like: `new Alert(elem, config)`?

If we streamline `_getConfig`, it can be omitted on `getOrCreateInstance`  I suppose

Comment (rohit2sharma95) [reply js/src/carousel.js:222]: Yes @GeoSot, I am also pointing the same thing. You can remove `typeof config === 'object' ? config : {}` from the `_getConfig` method IMO. :thinking:

Comment (GeoSot) [reply js/src/carousel.js:222]: and where are we  going to check it? :thinking:

Comment (rohit2sharma95) [reply js/src/carousel.js:222]: Okay NVM, if you are checking the configuration for the initialization of the plugin then it's good.

**Comment by rohit2sharma95** [comment]
This new method should be documented IMO.
/CC @XhmikosR

File: js/src/popover.js [hunk 8]
```diff
@@ -146,13 +145,7 @@ class Popover extends Tooltip {
 
   static jQueryInterface(config) {
     return this.each(function () {
-      let data = Data.get(this, DATA_KEY)
-      const _config = typeof config === 'object' ? config : null
-
-      if (!data) {
-        data = new Popover(this, _config)
-        Data.set(this, DATA_KEY, data)
```
Comment (rohit2sharma95) [inline js/src/popover.js:154]: Why this `Data.set(this, DATA_KEY, data)` was required before? :confused:

File: js/src/base-component.js [hunk 9]
```diff
@@ -61,6 +61,12 @@ class BaseComponent {
     return Data.get(element, this.DATA_KEY)
   }
 
+  static getOrCreateInstance(element, config = {}) {
+    const _config = typeof config === 'object' ? config : null
+
+    return this.getInstance(element) || new this(element, _config)
+  }
+
```
Comment (rohit2sharma95) [inline js/src/base-component.js:69]:
```suggestion
  static getOrCreateInstance(element, config = {}) {
    return this.getInstance(element) || new this(element, typeof config === 'object' ? config : null)
  }
```

To prevent the check if the instance already exists.

File: js/src/popover.js (hunk 8 above)
Comment (GeoSot) [reply js/src/popover.js:154]: I really feel was a left over. The component initialize its self using the base-component constructor.
There the element is saved on js/src/base-component.js:34
```js
 Data.set(this._element, this.constructor.DATA_KEY, this)
```
line 154 seems like a duplication which re-saves the same thing

**Comment by GeoSot** [comment]
I have added the proper documentation. So do we need something more in order to proceed?

File: js/src/offcanvas.js [hunk 10]
```diff
@@ -256,13 +255,12 @@ EventHandler.on(document, EVENT_CLICK_DATA_API, SELECTOR_DATA_TOGGLE, function (
...
 
-  const data = Data.get(target, DATA_KEY) || new Offcanvas(target)
-
+  const data = Offcanvas.getOrCreateInstance(target)
   data.toggle(this)
 })
 
 EventHandler.on(window, EVENT_LOAD_DATA_API, () => {
-  SelectorEngine.find(OPEN_SELECTOR).forEach(el => (Data.get(el, DATA_KEY) || new Offcanvas(el)).show())
+  SelectorEngine.find(OPEN_SELECTOR).forEach(el => (Offcanvas.getOrCreateInstance(el)).show())
```
Comment (alpadev) [inline js/src/offcanvas.js:263]: I think you can omit the brackets here.

File: js/tests/unit/alert.spec.js [hunk 11]
```diff
@@ -207,4 +207,26 @@ describe('Alert', () => {
...
+
+  describe('getOrCreateInstance', () => {
+    it('should return alert instance', () => {
+      fixtureEl.innerHTML = '<div></div>'
+
+      const div = fixtureEl.querySelector('div')
+      const alert = new Alert(div)
+
+      expect(Alert.getOrCreateInstance(div)).toEqual(alert)
+      expect(Alert.getInstance(div)).toEqual(Alert.getOrCreateInstance(div, {}))
```
Comment (alpadev) [inline js/tests/unit/alert.spec.js:219]: Isn't the second expect kind of redundant? If `getOrCreateInstance` equals `alert` already does it makes sense to test `getInstance` separately as this is used by `getOrCreateInstance`?

File: js/tests/unit/offcanvas.spec.js [hunk 12]
```diff
@@ -620,38 +620,6 @@ describe('Offcanvas', () => {
...
-
-      jQueryMock.fn.offcanvas = Offcanvas.jQueryInterface
-      jQueryMock.elements = [div]
-
-      try {
-        jQueryMock.fn.offcanvas.call(jQueryMock, action)
-      } catch (error) {
-        expect(error.message).toEqual(`No method named "${action}"`)
-      }
-    })
```
Comment (alpadev) [inline js/tests/unit/offcanvas.spec.js:653]: Seems to be unrelated change, no?

File: site/content/docs/5.0/components/alerts.md [hunk 13]
```diff
@@ -209,6 +209,15 @@ This makes an alert listen for click events on descendant elements which have th
         Static method which allows you to get the alert instance associated to a DOM element, you can use it like this: <code>bootstrap.Alert.getInstance(alert)</code>
       </td>
     </tr>
+    <tr>
+      <td>
+        <code>getOrCreateInstance</code>
+      </td>
+      <td>
+        Static method which returns an alert instance associated to a DOM element or create a new one in case it wasn't initialised.
+        You can use it like this: <code>bootstrap.Alert.getOrCreateInstance(alert)</code>
```
Comment (alpadev) [inline site/content/docs/5.0/components/alerts.md:218]: You're using `element` as the parameter in the other docs, but here you're using `alert`. Maybe stick to one way.

File: site/content/docs/5.0/components/list-group.md [hunk 14]
```diff
@@ -484,6 +484,15 @@ var triggerEl = document.querySelector('#trigger')
 var tab = bootstrap.Tab.getInstance(triggerEl) // Returns a Bootstrap tab instance
 ```
 
+#### getOrCreateInstance
+
+*Static* method which allows you to get the tab instance associated with a DOM element, or create a new one in case it wasn't initialised
+
+```js
+var triggerEl = document.querySelector('#trigger')
+var tab = bootstrap.Tab.getOrCreateInstance(triggerEl) // Returns a Bootstrap tab instance
```
Comment (alpadev) [inline site/content/docs/5.0/components/list-group.md:493]: The comment is the same as with `getInstance`. Maybe it makes sense to reflect the creational behaviour here too.

File: js/tests/unit/offcanvas.spec.js (hunk 12 above)
Comment (GeoSot) [reply js/tests/unit/offcanvas.spec.js:653]: Yes, however is duplicated :weary:

File: js/tests/unit/alert.spec.js (hunk 11 above)
Comment (GeoSot) [reply js/tests/unit/alert.spec.js:219]: I want to show this usage, but can omitted too

**Review by alpadev — APPROVED** [review]
LGTM, from my side.

File: js/tests/unit/base-component.spec.js [hunk 15]
```diff
@@ -112,5 +112,22 @@ describe('Base Component', () => {
...
+    describe('getOrCreateInstance', () => {
+      it('should return an instance', () => {
+        createInstance()
+
+        expect(DummyClass.getOrCreateInstance(element)).toEqual(instance)
+        expect(DummyClass.getInstance(element)).toEqual(DummyClass.getOrCreateInstance(element, {}))
+        expect(DummyClass.getOrCreateInstance(element)).toBeInstanceOf(DummyClass)
+      })
+
+      it('should return new instance when there is no alert instance', () => {
```
Comment (rohit2sharma95) [inline js/tests/unit/base-component.spec.js:124]: ~~alert~~

File: js/tests/unit/tooltip.spec.js [hunk 16]
```diff
@@ -3,7 +3,7 @@ import EventHandler from '../../src/dom/event-handler'
 import { noop } from '../../src/util/index'
 
 /** Test helpers */
-import { getFixture, clearFixture, jQueryMock, createEvent } from '../helpers/fixture'
+import { clearFixture, createEvent, getFixture, jQueryMock } from '../helpers/fixture'
```
Comment (rohit2sharma95) [inline js/tests/unit/tooltip.spec.js:6]: You must be using any code formatter in your IDE :sweat_smile:

File: js/src/offcanvas.js [hunk 17]
```diff
@@ -256,14 +255,13 @@ EventHandler.on(document, EVENT_CLICK_DATA_API, SELECTOR_DATA_TOGGLE, function (
...
+  const data = Offcanvas.getOrCreateInstance(target)
   data.toggle(this)
 })
 
-EventHandler.on(window, EVENT_LOAD_DATA_API, () => {
-  SelectorEngine.find(OPEN_SELECTOR).forEach(el => (Data.get(el, DATA_KEY) || new Offcanvas(el)).show())
-})
+EventHandler.on(window, EVENT_LOAD_DATA_API, () =>
+  SelectorEngine.find(OPEN_SELECTOR).forEach(el => (Offcanvas.getOrCreateInstance(el)).show())
+)
```
Comment (rohit2sharma95) [inline js/src/offcanvas.js:264]:
```suggestion
EventHandler.on(window, EVENT_LOAD_DATA_API, () =>
  SelectorEngine.find(OPEN_SELECTOR).forEach(el => Offcanvas.getOrCreateInstance(el).show())
)
```

**Review by rohit2sharma95 — APPROVED** [review]
/CC @mdo for the documentation, if required

File: js/tests/unit/tooltip.spec.js (hunk 16 above)
Comment (GeoSot) [reply js/tests/unit/tooltip.spec.js:6]: HAHAHA jetBrains like them sorted :yum:
